File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> tracking;
        
        int res=0, first=0;
        
        for(int i=0; i<s.size(); i++){
            if(tracking[s[i]]){
                first = max(first, tracking[s[i]]);
            }
            tracking[s[i]]=i+1;

            res = max(res, i-first+1);
            // cout<<res<<" "<<first<<" "<<i<<endl;
        }
        return res;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> present = new HashMap<>();
        
        int left=0, res =0;
        
        for(int right=0; right<s.length(); right++){
            char c = s.charAt(right);
            
            if(present.containsKey(c) ){
                left = Math.max(left, present.get(c)+1);
            }
            present.put(c, right);
             res = Math.max(res, right-left+1);
        }
        return res;
    }
}